app/src/main/java/org/dhis2/data/service/SyncPresenterImpl.kt

```diff
@@ -24,23 +22,45 @@ import org.dhis2.data.service.workManager.WorkManagerController
 import org.dhis2.data.service.workManager.WorkerItem
 import org.dhis2.data.service.workManager.WorkerType
 import org.dhis2.utils.DateUtils
+import org.dhis2.utils.analytics.AnalyticsHelper
 import org.hisp.dhis.android.core.D2
 import org.hisp.dhis.android.core.arch.call.D2Progress
 import org.hisp.dhis.android.core.common.State
 import org.hisp.dhis.android.core.imports.TrackerImportConflict
 import org.hisp.dhis.android.core.program.ProgramType
+import org.hisp.dhis.android.core.settings.GeneralSettings
+import org.hisp.dhis.android.core.settings.LimitScope
+import org.hisp.dhis.android.core.settings.ProgramSettings
+import org.hisp.dhis.android.core.systeminfo.DHISVersion
 import timber.log.Timber
+import java.util.ArrayList
+import java.util.Calendar
+import kotlin.math.ceil
 
 class SyncPresenterImpl(
     private val d2: D2,
     private val preferences: PreferenceProvider,
-    private val workManagerController: WorkManagerController
+    private val workManagerController: WorkManagerController,
+    private val analyticsHelper: AnalyticsHelper
 ) : SyncPresenter {
 
     override fun syncAndDownloadEvents() {
-        val eventLimit = preferences.getInt(EVENT_MAX, EVENT_MAX_DEFAULT)
-        val limitByOU = preferences.getBoolean(LIMIT_BY_ORG_UNIT, false)
-        val limitByProgram = preferences.getBoolean(LIMIT_BY_PROGRAM, false)
+        val programSettings = getProgramSetting()
+        val globalProgramSettings = programSettings.globalSettings()
+
+        val eventLimit = globalProgramSettings?.eventsDownload() ?: preferences.getInt(
+            EVENT_MAX,
+            EVENT_MAX_DEFAULT
+        )
+
+        val limitByOU = globalProgramSettings?.settingDownload()?.let {
+            it == LimitScope.PER_ORG_UNIT || it == LimitScope.PER_OU_AND_PROGRAM
+        } ?: preferences.getBoolean(LIMIT_BY_ORG_UNIT, false)
+
+        val limitByProgram = globalProgramSettings?.settingDownload()?.let {
+            it == LimitScope.PER_PROGRAM || it == LimitScope.PER_OU_AND_PROGRAM
+        } ?: preferences.getBoolean(LIMIT_BY_PROGRAM, false)
```

Comment:
this could be a very valuable unit test doing verify's for 
         Completable.fromObservable(
                    d2.eventModule()
                        .eventDownloader()
                        .limit(eventLimit)
                        .limitByOrgunit(limitByOU)
                        .limitByProgram(limitByProgram)
                        .download()

Do you think this could be possible?